File: Sample Code/ObjcSampleCode/DJISdkDemo/Demo/DJIMapView/DJIAnnotationViews/DJIAnnotation.m
//
//  DJIAnnotation.m
//  Phantom3
//
//  Created by Jayce Yang on 14-4-10.
//  Copyright (c) 2014年 Jerome.zhang. All rights reserved.
//

#import "DJIAnnotation.h"
#import "DJIAnnotationView.h"

@interface DJIAnnotation ()

@end

@implementation DJIAnnotation

- (void)dealloc
{
    if (_backGroundAnnotation && _mapView) {
        if ([NSThread isMainThread]) {
           [self.mapView removeAnnotation:_backGroundAnnotation];
        } else {
            dispatch_sync(dispatch_get_main_queue(), ^{
                [self.mapView removeAnnotation:self.backGroundAnnotation];
            });
        }
        
    }
}

- (id)init
{
    self = [super init];
    if (self) {
        // Initialization code
        
        self.index = 0;
    }
    return self;
}

//- (void)dealloc
//{
//    NSLog();
//}

+ (instancetype)annotationWithCoordinate:(CLLocationCoordinate2D)coordinate
{
    DJIAnnotation *annotation = [[self alloc] init];
    annotation.coordinate = coordinate;
    return annotation;
}

- (void)setCoordinate:(CLLocationCoordinate2D)coordinate
{
    _coordinate = coordinate;
    if (_backGroundAnnotation) {
        [_backGroundAnnotation setCoordinate:coordinate];
    }
}

- (void)setCoordinate:(CLLocationCoordinate2D)coordinate animation:(bool)animation
{
    _coordinate = coordinate;
    if (_backGroundAnnotation) {
        [_backGroundAnnotation setCoordinate:coordinate animation:animation];
    }
}

- (void)createBackgroundAnnotation
{
    DJIAnnotation *annotation = [DJIAnnotation annotationWithCoordinate:_coordinate];
    annotation.isBackgroundAnnotation = YES;
    
    if (_mapView) {
        [_mapView addAnnotation:annotation];
    }
    
    _backGroundAnnotation = annotation;
}
@end

File: Sample Code/SwiftSampleCode/DJISDKSwiftDemo/Missions/DJIMapView/DJIAnnotationViews/DJIAnnotation.swift
//
//  DJIAnnotation.swift
//  DJISDKSwiftDemo
//
import Foundation
import MapKit
import CoreLocation
class DJIAnnotation: NSObject, MKAnnotation {
    /**
     *  Index of the annotation
     */
    var index: String? = nil
    weak var annotationView: MKAnnotationView? = nil
    var backGroundAnnotation: DJIAnnotation? = nil
    weak var mapView: MKMapView?=nil
    var isBackgroundAnnotation: Bool?

    var _coordinate:CLLocationCoordinate2D
    dynamic var coordinate: CLLocationCoordinate2D{
        get {
            return _coordinate
        }
        set (coordinate){
            _coordinate = coordinate
            if backGroundAnnotation != nil  {
                backGroundAnnotation!.coordinate = coordinate
            }
        }
    }



    init(coordinate: CLLocationCoordinate2D) {
        _coordinate = coordinate
        super.init()
    }

    func createBackgroundAnnotation() {
        let annotation: DJIAnnotation = DJIAnnotation(coordinate:_coordinate)
        annotation.isBackgroundAnnotation = true
        if mapView != nil {
            mapView!.addAnnotation(annotation)
        }
        self.backGroundAnnotation = annotation
    }

    func setCoordinate(coordinate: CLLocationCoordinate2D, animation: Bool) {
        self.coordinate = coordinate
        if backGroundAnnotation != nil {
            backGroundAnnotation!.setCoordinate(coordinate, animation: animation)
        }
    }

    deinit {
        if backGroundAnnotation != nil && mapView != nil {
            if NSThread.isMainThread() {
                mapView!.removeAnnotation(backGroundAnnotation!)
            }
            else {
                dispatch_sync(dispatch_get_main_queue(), {[weak self] () -> Void in
                    self?.mapView!.removeAnnotation(self!.backGroundAnnotation!)
                })
            }
        }
    }

}
